Index: Bug in even and odd

Hi again,


I'd like to report some bugs in 'even' and 'odd'. (Clash, version 1.0.1 (using clash-lib, version: 1.0.1))

    -- even :: Integral a => a -> Bool

    bug1 = even (0 :: Index 1)
    bug2 = even (0 :: Index 2)
    bug3 = even (1 :: Index 2)

Now bug1 gives

*** Exception: X: Clash.Sized.Index: result 2 is out of bounds: [0..0]
CallStack (from HasCallStack):
  errorX, called at src/Clash/Sized/Internal/Index.hs:255:13 in clash-prelude-1.0.1-BorHjli5pMX3STUvZInnYc:Clash.Sized.Internal.Index
  fromInteger_INLINE, called at src/Clash/Sized/Internal/Index.hs:249:16 in clash-prelude-1.0.1-BorHjli5pMX3STUvZInnYc:Clash.Sized.Internal.Index

You get similar errors for 'odd'.

Here's a possible fix:

    odd', even' :: (KnownNat n) => Index (n+1) -> Bool                         
    odd' = bitToBool . lsb . pack                                     
    even' = not . odd'

  * These errors are __extremely__ difficult to find because the error messages thrown do not refer to sections in the user's source file. If the error is caused by a statement/function call in a deeply nested where clause, you can only find the error by making the deeply nested statements top-level functions and play with these functions until you've found the cause. It can take very long before you know which statement is the cause...
  * I get similar errors like these a lot. They're usually caused by errors on my behalf. Finding them can take very long.
  * Is there any way you can introduce a flag which forces error messages to pin-point the cause of the error to a location in the user's source file?

TIA for your time.

Regards,


Paddy

And for the "I wanna write a Index constant and have it checked at compile time"-problem you could also do:

```haskell
liftIndex :: KnownNat n => Index n -> Q (TExp (Index n))
liftIndex i = case hasX i of
  Left err -> fail $ "liftIndex failed: " <> err
  Right _ -> TExp <$> lift i

test :: Index 3
test = $$(liftIndex 10)
```

```
testlit.hs:5:11: error:
    • liftIndex failed: X: Clash.Sized.Index: result 10 is out of bounds: [0..2]
CallStack (from HasCallStack):
  errorX, called at src/Clash/Sized/Internal/Index.hs:255:13 in clash-prelude-1.1.0-inplace:Clash.Sized.Internal.Index
  fromInteger_INLINE, called at src/Clash/Sized/Internal/Index.hs:249:16 in clash-prelude-1.1.0-inplace:Clash.Sized.Internal.Index
    • In the Template Haskell splice $$(liftIndex 10)
      In the expression: $$(liftIndex 10)
      In an equation for ‘test’: test = $$(liftIndex 10)
```

> And for the "I wanna write a Index constant and have it checked at compile time"-problem you could also do:
> 
> ```haskell
> module LiftIndex where
> import Clash.Prelude
> import Language.Haskell.TH.Syntax
> liftIndex :: KnownNat n => Index n -> Q (TExp (Index n))
> liftIndex i = case hasX i of
>   Left err -> fail $ "liftIndex failed: " <> err
>   Right _ -> TExp <$> lift i
> 
> -- put this in a different file:
> test :: Index 3
> test = $$(liftIndex 10)
> ```
> 
> ```
> testlit.hs:5:11: error:
>     • liftIndex failed: X: Clash.Sized.Index: result 10 is out of bounds: [0..2]
> CallStack (from HasCallStack):
>   errorX, called at src/Clash/Sized/Internal/Index.hs:255:13 in clash-prelude-1.1.0-inplace:Clash.Sized.Internal.Index
>   fromInteger_INLINE, called at src/Clash/Sized/Internal/Index.hs:249:16 in clash-prelude-1.1.0-inplace:Clash.Sized.Internal.Index
>     • In the Template Haskell splice $$(liftIndex 10)
>       In the expression: $$(liftIndex 10)
>       In an equation for ‘test’: test = $$(liftIndex 10)
> ```

Thanks for the suggestion Rowan. Requires a lot of work, though. 

Regards,


Paddy

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -2,6 +2,9 @@
 
 ## 1.1.0
 * New features (API):
+  * `Clash.Class.Parity` type class replaces Prelude `odd` and `even` functions due
+    to assumptions that don't hold for Clash specific numerical types, see
+    [#970](https://github.com/clash-lang/clash-compiler/pull/970).
   * `NFDataX.ensureSpine`, see [#748](https://github.com/clash-lang/clash-compiler/pull/803)
   * `makeTopEntity` Template Haskell function for generating TopEntity annotations
     intended to cover the majority of use cases. Generation failures should either

--- a/clash-prelude/clash-prelude.cabal
+++ b/clash-prelude/clash-prelude.cabal
@@ -150,6 +150,7 @@ Library
                       Clash.Class.HasDomain.CodeGen
                       Clash.Class.HasDomain.Common
                       Clash.Class.Num
+                      Clash.Class.Parity
                       Clash.Class.Resize
 
                       Clash.Clocks

--- a/clash-prelude/src/Clash/Class/Parity.hs
+++ b/clash-prelude/src/Clash/Class/Parity.hs
@@ -1,0 +1,125 @@
+{-|
+Copyright  :  (C) 2019, QBayLogic
+License    :  BSD2 (see the file LICENSE)
+Maintainer :  Christiaan Baaij <christiaan.baaij@gmail.com>
+-}
+
+{-# LANGUAGE CPP                  #-}
+{-# LANGUAGE DefaultSignatures    #-}
+{-# LANGUAGE FlexibleContexts     #-}
+{-# LANGUAGE ScopedTypeVariables  #-}
+{-# LANGUAGE TypeApplications     #-}
+{-# LANGUAGE MagicHash            #-}
+
+{-# OPTIONS_GHC -fplugin=GHC.TypeLits.Extra.Solver #-}
+{-# OPTIONS_GHC -fplugin=GHC.TypeLits.Normalise #-}
+{-# OPTIONS_GHC -fplugin=GHC.TypeLits.KnownNat.Solver #-}
+{-# OPTIONS_HADDOCK show-extensions #-}
+
+#include "MachDeps.h"
+
+module Clash.Class.Parity
+  ( Parity (..) )
+where
+
+import Prelude hiding                 (even, odd)
+
+import Data.Int
+import Data.Word
+import Foreign.C.Types                (CUShort)
+import GHC.TypeLits                   (KnownNat)
+
+import Clash.Class.BitPack            (pack)
+import Clash.Sized.Internal.BitVector (BitVector, high, low, lsb#)
+import Clash.Promoted.Nat             (SNat(..), snatToNum)
+
+{- $setup
+>>> import Clash.Prelude
+-}
+
+-- | Determine whether value is odd or even
+class Parity a where
+  -- | Check if value is even
+  --
+  -- >>> even (4 :: Unsigned 4)
+  -- True
+  even :: a -> Bool
+  even = not . odd
+
+  -- | Check if value is odd
+  --
+  -- >>> odd (4 :: Unsigned 4)
+  -- False
+  odd :: a -> Bool
+  odd = not . even
+  {-# MINIMAL even | odd #-}
+
+instance Parity Integer
+  where
+    even a = a `mod` 2 == 0
+    odd a = a `mod` 2 == 1
+
+instance KnownNat n => Parity (BitVector n) where
+  even a =
+    case snatToNum @Integer (SNat @n) of
+      0 -> True
+      _ -> (==low) $ lsb# a
+  odd a =
+    case snatToNum @Integer (SNat @n) of
+      0 -> False
+      _ -> (==high) $ lsb# a
+
+instance Parity Bool where
+  even = even . pack
+  odd = odd . pack
+
+instance Parity CUShort where
+  even = even . pack
+  odd = odd . pack
+
+instance Parity Word where
+  even = even . pack
+  odd = odd . pack
+
+instance Parity Word8 where
+  even = even . pack
+  odd = odd . pack
+
+instance Parity Word16 where
+  even = even . pack
+  odd = odd . pack
+
+instance Parity Word32 where
+  even = even . pack
+  odd = odd . pack
+
+
+#if WORD_SIZE_IN_BITS >= 64
+instance Parity Word64 where
+  even = even . pack
+  odd = odd . pack
+#endif
+
+instance Parity Int where
+  even = even . pack
+  odd = odd . pack
+
+instance Parity Int8 where
+  even = even . pack
+  odd = odd . pack
+
+instance Parity Int16 where
+  even = even . pack
+  odd = odd . pack
+
+instance Parity Int32 where
+  even = even . pack
+  odd = odd . pack
+
+#if WORD_SIZE_IN_BITS >= 64
+instance Parity Int64 where
+  even = even . pack
+  odd = odd . pack
+#endif
+
+

--- a/clash-prelude/src/Clash/HaskellPrelude.hs
+++ b/clash-prelude/src/Clash/HaskellPrelude.hs
@@ -4,8 +4,8 @@
   Maintainer  :  QBayLogic B.V <devops@qbaylogic.com>
 
 "Clash.HaskellPrelude" re-exports most of the Haskell "Prelude" with the exception of
-the following: (++), (!!), concat, drop, foldl, foldl1, foldr, foldr1, head,
-init, iterate, last, length, map, repeat, replicate, reverse, scanl, scanr,
+the following: (++), (!!), concat, drop, even, foldl, foldl1, foldr, foldr1, head,
+init, iterate, last, length, map, odd, repeat, replicate, reverse, scanl, scanr,
 splitAt, tail, take, unzip, unzip3, zip, zip3, zipWith, zipWith3.
 -}
 
@@ -18,6 +18,6 @@ module Clash.HaskellPrelude
 where
 
 import Prelude hiding
-  ((++), (!!), concat, concatMap, drop, foldl, foldl1, foldr, foldr1, head, init,
-   iterate, last, length, map, repeat, replicate, reverse, scanl, scanr, splitAt,
+  ((++), (!!), concat, concatMap, drop, even, foldl, foldl1, foldr, foldr1, head, init,
+   iterate, last, length, map, odd, repeat, replicate, reverse, scanl, scanr, splitAt,
    tail, take, unzip, unzip3, zip, zip3, zipWith, zipWith3, undefined, (^))

--- a/clash-prelude/src/Clash/Prelude.hs
+++ b/clash-prelude/src/Clash/Prelude.hs
@@ -139,6 +139,7 @@ module Clash.Prelude
   , module Clash.Class.BitPack
   , module Clash.Class.Exp
   , module Clash.Class.Num
+  , module Clash.Class.Parity
   , module Clash.Class.Resize
     -- *** Other
   , module Control.Applicative
@@ -173,6 +174,7 @@ import           Clash.Class.AutoReg         (AutoReg, deriveAutoReg)
 import           Clash.Class.BitPack
 import           Clash.Class.Exp
 import           Clash.Class.Num
+import           Clash.Class.Parity
 import           Clash.Class.Resize
 import qualified Clash.Explicit.Prelude      as E
 import           Clash.Hidden
@@ -212,12 +214,13 @@ import           Clash.XException
 
 {- $hiding
 "Clash.Prelude" re-exports most of the Haskell "Prelude" with the exception of
-the following: (++), (!!), concat, drop, foldl, foldl1, foldr, foldr1, head,
-init, iterate, last, length, map, repeat, replicate, reverse, scanl, scanr,
+the following: (++), (!!), concat, drop, even, foldl, foldl1, foldr, foldr1, head,
+init, iterate, last, length, map, odd, repeat, replicate, reverse, scanl, scanr,
 splitAt, tail, take, unzip, unzip3, zip, zip3, zipWith, zipWith3.
 
 It instead exports the identically named functions defined in terms of
-'Clash.Sized.Vector.Vec' at "Clash.Sized.Vector".
+'Clash.Sized.Vector.Vec' at "Clash.Sized.Vector". For the 'odd' end 'even'
+function a type class called Parity is available at 'Clash.Class.Parity'.
 -}
 
 

--- a/clash-prelude/src/Clash/Sized/Internal/Index.hs
+++ b/clash-prelude/src/Clash/Sized/Internal/Index.hs
@@ -77,6 +77,8 @@ module Clash.Sized.Internal.Index
   )
 where
 
+import Prelude hiding             (even, odd)
+
 import Control.DeepSeq            (NFData (..))
 import Data.Bits                  (Bits (..), FiniteBits (..))
 import Data.Data                  (Data)
@@ -98,6 +100,7 @@ import Test.QuickCheck.Arbitrary  (Arbitrary (..), CoArbitrary (..),
 import Clash.Class.BitPack        (BitPack (..), packXWith)
 import Clash.Class.Num            (ExtendingNum (..), SaturatingNum (..),
                                    SaturationMode (..))
+import Clash.Class.Parity         (Parity (..))
 import Clash.Class.Resize         (Resize (..))
 import Clash.Prelude.BitIndex     (replaceBit)
 import {-# SOURCE #-} Clash.Sized.Internal.BitVector (BitVector (BV), high, low, undefError)
@@ -351,6 +354,10 @@ quot#,rem# :: Index n -> Index n -> Index n
 {-# NOINLINE toInteger# #-}
 toInteger# :: Index n -> Integer
 toInteger# (I n) = n
+
+instance (KnownNat n, 1 <= n) => Parity (Index n) where
+  even = even . pack
+  odd = odd . pack
 
 instance (KnownNat n, 1 <= n) => Bits (Index n) where
   a .&. b           = unpack# $ BV.and# (pack# a) (pack# b)

--- a/clash-prelude/src/Clash/Sized/Internal/Signed.hs
+++ b/clash-prelude/src/Clash/Sized/Internal/Signed.hs
@@ -85,6 +85,8 @@ module Clash.Sized.Internal.Signed
   )
 where
 
+import Prelude hiding                 (odd, even)
+
 import Control.DeepSeq                (NFData (..))
 import Control.Lens                   (Index, Ixed (..), IxValue)
 import Data.Bits                      (Bits (..), FiniteBits (..))
@@ -104,6 +106,7 @@ import Test.QuickCheck.Arbitrary      (Arbitrary (..), CoArbitrary (..),
 import Clash.Class.BitPack            (BitPack (..), packXWith)
 import Clash.Class.Num                (ExtendingNum (..), SaturatingNum (..),
                                        SaturationMode (..))
+import Clash.Class.Parity             (Parity (..))
 import Clash.Class.Resize             (Resize (..))
 import Clash.Prelude.BitIndex         ((!), msb, replaceBit, split)
 import Clash.Prelude.BitReduction     (reduceAnd, reduceOr)
@@ -355,6 +358,10 @@ mod# (S a) (S b) = S (a `mod` b)
 {-# NOINLINE toInteger# #-}
 toInteger# :: Signed n -> Integer
 toInteger# (S n) = n
+
+instance KnownNat n => Parity (Signed n) where
+  even = even . pack
+  odd = odd . pack
 
 instance KnownNat n => Bits (Signed n) where
   (.&.)             = and#

--- a/clash-prelude/src/Clash/Sized/Internal/Unsigned.hs
+++ b/clash-prelude/src/Clash/Sized/Internal/Unsigned.hs
@@ -77,6 +77,8 @@ module Clash.Sized.Internal.Unsigned
   )
 where
 
+import Prelude hiding                 (even, odd)
+
 import Control.DeepSeq                (NFData (..))
 import Control.Lens                   (Index, Ixed (..), IxValue)
 import Data.Bits                      (Bits (..), FiniteBits (..))
@@ -97,6 +99,7 @@ import Test.QuickCheck.Arbitrary      (Arbitrary (..), CoArbitrary (..),
 import Clash.Class.BitPack            (BitPack (..), packXWith)
 import Clash.Class.Num                (ExtendingNum (..), SaturatingNum (..),
                                        SaturationMode (..))
+import Clash.Class.Parity             (Parity (..))
 import Clash.Class.Resize             (Resize (..))
 import Clash.Prelude.BitIndex         ((!), msb, replaceBit, split)
 import Clash.Prelude.BitReduction     (reduceOr)
@@ -332,6 +335,10 @@ rem# (U i) (U j) = U (i `rem` j)
 {-# NOINLINE toInteger# #-}
 toInteger# :: Unsigned n -> Integer
 toInteger# (U i) = i
+
+instance KnownNat n => Parity (Unsigned n) where
+  even = even . pack
+  odd = odd . pack
 
 instance KnownNat n => Bits (Unsigned n) where
   (.&.)             = and#

--- a/tests/shouldwork/Numbers/NumConstantFolding_1.hs
+++ b/tests/shouldwork/Numbers/NumConstantFolding_1.hs
@@ -65,6 +65,21 @@ cIntegral = (rQuot,rRem,rDiv,rMod,rQuotRem,rDivMod)
     rDivMod  = divMod  @n (lit 22155) (lit 46)
     -- toInteger?
 
+cParity :: forall n. (Num n, Parity n) => _
+cParity = (rEven, rOdd)
+  where
+    rEven = even @n (lit 22290)
+    rOdd  = odd  @n (lit 22291)
+
+cParityBS0 :: forall n. (Num n, Parity n, BitPack n, 0 ~ (BitSize n)) => _
+cParityBS0 = (rEven, rOdd)
+  where
+    f :: Int -> Bool -> Int
+    f a = \x -> case x of {True -> 0; False -> a}
+
+    rEven = f 22302 $ even @n (lit 22300)
+    rOdd  = f 22303 $ odd  @n (lit 22301) == False
+
 cBitsNoPopCount :: forall n. (Num n, Bits n) => _
 -- the nested tuple here is so we can still 'show' the result
 cBitsNoPopCount = ((r00,r01,r02,r03,r04,r05,r06,r07,r08,r09,r10),r11,r12,r13,r14,r15,r16,r17,r18,r19)
@@ -196,6 +211,8 @@ tUnsigned16
     , cEq             @(Unsigned 16)
     , cOrd            @(Unsigned 16)
     , cIntegral       @(Unsigned 16)
+    , cParity         @(Unsigned 16)
+    , cParityBS0      @(Unsigned 0 )
     , cBits           @(Unsigned 16)
     -- , cFiniteBits  @(Unsigned 16) -- broken
     , csClashSpecific @(Unsigned 16)
@@ -207,6 +224,8 @@ tSigned16
     , cEq             @(Signed 16)
     , cOrd            @(Signed 16)
     , cIntegral       @(Signed 16)
+    , cParity         @(Signed 16)
+    , cParityBS0      @(Signed 0 )
     , cBits           @(Signed 16)
     -- , cFiniteBits  @(Signed 16) -- broken
     , csClashSpecific @(Signed 16)
@@ -218,6 +237,8 @@ tBitVector16
     , cEq             @(BitVector 16)
     , cOrd            @(BitVector 16)
     , cIntegral       @(BitVector 16)
+    , cParity         @(BitVector 16)
+    , cParityBS0      @(BitVector 0 )
     , cBits           @(BitVector 16)
     -- , cFiniteBits  @(BitVector 16) -- broken
     , csClashSpecific @(BitVector 16)
@@ -229,12 +250,15 @@ tIndex
     , cEq             @(Index 50000)
     , cOrd            @(Index 50000)
     , cIntegral       @(Index 50000)
+    , cParity         @(Index 50000)
+    , cParityBS0      @(Index 1    )
     , cBits           @(Index 65536)
     -- , cFiniteBits     @(Index 50000) -- broken
     , cIndex1 -- ensure special case for index 1 is verified
     , csClashSpecific @(Index 50000)
     , cResize         @(Index 50000)
     )
+
 tSFixed
   = ( cNum            @(SFixed 16 0)
     , cEq             @(SFixed 16 0)

